### PR TITLE
Add Get/Update for api/v1/user/applications/oauth2

### PR DESCRIPTION
**Pull request by dmolik**
Add API methods for getting and updating user oauth2 applications.

### CODE REVIEW

File: routers/api/v1/user/app.go [hunk 1]
```diff
@@ -231,3 +231,87 @@ func DeleteOauth2Application(ctx *context.APIContext) {
...
+	if err != nil {
+		if models.IsErrOauthClientIDInvalid(err) {
+			ctx.NotFound()
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetOauth2ApplicationByID", err)
+		}
+		return
+	}
+
+	ctx.JSON(http.StatusOK, convert.ToOAuth2Application(app))
```
Comment (6543) [inline routers/api/v1/user/app.go:265]: Pleace delete `ClientSecret` before responding

Comment (dmolik) [reply routers/api/v1/user/app.go:265]: Done!

**Comment by dmolik** [comment]
Had some fixups with registering the new endpoints and passing the tests.

**Comment by dmolik** [comment]
looks like an unrelated postgres logging error

**Comment by 6543** [comment]
@dmolik pleace read the Contributing Guidline for API: https://github.com/go-gitea/gitea/blob/master/CONTRIBUTING.md#api-v1

File: integrations/api_oauth2_apps_test.go [hunk 2]
```diff
@@ -83,9 +85,6 @@ func testAPIDeleteOAuth2Application(t *testing.T) {
 	oldApp := models.AssertExistsAndLoadBean(t, &models.OAuth2Application{
 		UID:  user.ID,
 		Name: "test-app-1",
-		RedirectURIs: []string{
```
Comment (6543) [inline integrations/api_oauth2_apps_test.go:86]: reason to remove this?

File: routers/api/v1/api.go [hunk 3]
```diff
@@ -580,7 +580,10 @@ func RegisterRoutes(m *macaron.Macaron) {
 				m.Combo("/oauth2").
 					Get(user.ListOauth2Applications).
 					Post(bind(api.CreateOAuth2ApplicationOptions{}), user.CreateOauth2Application)
-				m.Delete("/oauth2/:id", user.DeleteOauth2Application)
+				m.Combo("/oauth2/:id").
+					Delete(user.DeleteOauth2Application).
+					Post(bind(api.CreateOAuth2ApplicationOptions{}), user.UpdateOauth2Application).
```
Comment (6543) [inline routers/api/v1/api.go]: Wrong HTTP method

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by 6543** [comment]
@dmolik If you do not forcepush it's easier to review (pull get squash-merged at the end)

File: routers/api/v1/user/app.go [hunk 4]
```diff
@@ -231,3 +231,89 @@ func DeleteOauth2Application(ctx *context.APIContext) {
...
+	}
+
+	app.ClientSecret = ""
+
+	ctx.JSON(http.StatusOK, convert.ToOAuth2Application(app))
+}
+
+// UpdateOauth2Application update OAuth2 Application
+func UpdateOauth2Application(ctx *context.APIContext, data api.CreateOAuth2ApplicationOptions) {
+	// swagger:operation POST /user/applications/oauth2/{id} user userUpdateOAuth2Application
```
Comment (6543) [inline routers/api/v1/user/app.go]:
```suggestion
	// swagger:operation PATCH /user/applications/oauth2/{id} user userUpdateOAuth2Application
```

Comment (6543) [reply routers/api/v1/user/app.go]: and commit the result of `make generate-swagger`

File: integrations/api_oauth2_apps_test.go (hunk 2 above)
Comment (6543) [reply integrations/api_oauth2_apps_test.go:86]: ok because of `testAPIUpdateOAuth2Application` ... you can resolve this

Comment (dmolik) [reply integrations/api_oauth2_apps_test.go:86]: Yes exactly, suggestions?

File: routers/api/v1/api.go (hunk 3 above)
Comment (dmolik) [reply routers/api/v1/api.go]: yeap

File: routers/api/v1/user/app.go (hunk 4 above)
Comment (dmolik) [reply routers/api/v1/user/app.go]: :+1:

File: routers/api/v1/user/app.go [hunk 5]
```diff
@@ -231,3 +231,89 @@ func DeleteOauth2Application(ctx *context.APIContext) {
...
+		Name:         data.Name,
+		UserID:       ctx.User.ID,
+		ID:           appID,
+		RedirectURIs: data.RedirectURIs,
+	})
+	if err != nil {
+		ctx.Error(http.StatusBadRequest, "", "error updating oauth2 application")
+		return
+	}
+	app, err := models.GetOAuth2ApplicationByID(appID)
```
Comment (6543) [inline routers/api/v1/user/app.go:302]: I would move **GetOAuth2ApplicationByID** bevore ***UpdateOAuth2Application**

**Review by 6543 — COMMENTED** [review]
last nit

File: routers/api/v1/user/app.go (hunk 5 above)
Comment (dmolik) [reply routers/api/v1/user/app.go:302]: I'm concerned we would then return out of data information back to the client

Comment (6543) [reply routers/api/v1/user/app.go:302]: error handling is not the best ... we need a refactor there :(  ... I'm ok with this - but there should be a check if OAuth2App exist before Update :(

**resolfe** this for now. I'll send a refactor afterwards

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
Note for me: if merged, prepare **UpdateOAuth2Application** to return the app result back and handle errors

File: routers/api/v1/user/app.go (hunk 5 above)
Comment (dmolik) [reply routers/api/v1/user/app.go:302]: I think the proper fix here is to refactor the model to throw explicit errors, and to return the updated app, I'm okay with fixing it here, however I feel that it is out of scope of this PR.

Comment (6543) [reply routers/api/v1/user/app.go:302]: @dmolik leafe it like that - I'll take care of it later

**Review by lafriks — APPROVED** [review]

**Comment by dmolik** [comment]
Okie dokie, I went ahead and squashed the commits

**Comment by dmolik** [comment]
Looks like there were some mysql and mssql test failures, unrelated to this change.

**Review by 6543 — APPROVED** [review]
just push an empty commit to restart CI

**Comment by dmolik** [comment]
@6543 sweet the build passed! want to give it a once over?

**Comment by jolheiser** [comment]
Ping lgtm

**Comment by 6543** [comment]
created refactor #11034

**Pull request merged**